src/providers/ZapInitProvider/Zapper.ts

```diff
@@ -0,0 +1,490 @@
+
+    const getRateInQuoteSafe = this.zapData.abi.getRateInQuoteSafe;
+
+    if (!getRateInQuoteSafe) {
+      throw new Error('getRateInQuoteSafe not found in abi');
+    }
+
+    const client = createPublicClient({
+      chain: hyperevm,
+      transport: http(),
+    });
+
+    const abi: Abi = [getRateInQuoteSafe];
+    const contract = getContract({
+      address: this.getAbiAddress(getRateInQuoteSafe),
+      abi,
+      client,
+    });
+
+    const rate = await contract.read.getRateInQuoteSafe([token]);
+
+    if (!rate || typeof rate !== 'bigint') {
+      throw new Error('Failed to get rate');
+    }
+
+    const numerator = amount * 10n ** BigInt(decimals);
+    const denominator = rate;
+    return numerator / denominator;
+  };
+
+  // Override getSteps to add Hyperwave-specific steps
+  // @Note We should execute the steps in the correct order
+  // getSteps = (): string[] => {
+  //   return ['approve', 'deposit', 'transfer', 'customExtraStep'];
+  // };
+
+  getCommands = (): Record<string, ZapInstruction> => {
+    return hyperwaveCommands;
+  };
+}
+
+/**
+ * Factory function to create project-specific zapper instances.
+ * Each project can have its own custom contract instruction building logic.
+ *
+ * The general flow (raw calldata) remains in ZapInitProvider, while
+ * project-specific contract instructions are handled by the zapper classes.
+ *
+ * MODULAR FUNCTION SYSTEM:
+ * - Each project defines its steps via getSteps()
+ * - Each step is executed via a ZapInstruction function
+ * - Projects can override getSteps() to add/remove/reorder steps
+ * - Projects can override getCommands() to customize step behavior with different functions
+ *
+ * To add a new project:
+ * 1. Create a new class extending DefaultZapper
+ * 2. Override getSteps() to define your step sequence
+ * 3. Create project-specific ZapInstruction functions if needed
+ * 4. Override getCommands() to provide custom ZapInstruction functions
+ *
+ * Example:
+ * ```typescript
+ * case 'newproject':
+ *   return new NewProjectZapper(
+ *     sendCallsExtraParams.projectData,
+ *     sendCallsExtraParams.zapData,
+ *     sendCallsExtraParams.currentRoute,
+ *   );
+ * ```
+ *
+ * @param sendCallsExtraParams - Parameters containing project data and route information
+ * @returns A zapper instance configured for the specific project
+ */
+export const makeZapper = (sendCallsExtraParams: SendCallsExtraParams) => {
+  if (!sendCallsExtraParams.currentRoute) {
+    throw new Error('Current route is not set');
+  }
+
+  switch (sendCallsExtraParams.projectData.project) {
+    case 'hyperwave':
+      return new HyperwaveZapper(
+        sendCallsExtraParams.projectData,
+        sendCallsExtraParams.zapData,
+        sendCallsExtraParams.currentRoute,
+      );
+    default:
+      return new DefaultZapper(
+        sendCallsExtraParams.projectData,
+        sendCallsExtraParams.zapData,
+        sendCallsExtraParams.currentRoute,
+      );
+  }
+};
```

Comment:
If we're going this route, we have

`makeZapper` switches on the type to create a class (one level of indirection), that then expose different implementation of getCommands (one level of indirection), that returns different const values (another indirection).

You can get the same flexibility, while removing 2 layers, and dropping the class dispatches:

```suggestion
export const buildContractInstructions = (sendCallsExtraParams: SendCallsExtraParams) => {
  if (!sendCallsExtraParams.currentRoute) {
    throw new Error('Current route is not set');
  }

  switch (sendCallsExtraParams.projectData.project) {
    case 'hyperwave':
      return buildContractInstructions_(hyperwaveCommands, sendCallsExtraParams);
    default:
      return buildContractInstructions_(defaultCommands, sendCallsExtraParams);
  }
};
```